### ISSUE TITLE
Setting timeout for a keyword breaks formatting o log.html 

### ISSUE THREAD

**Issue by michalxkochel** [robotframework/robotframework#3859]
When I define timeout for a keyword, formatting of log.html is broken.
Logs which should be under keyword are grouped under keyword where timeout is defined.
See images bellow.

Logs without timeout
![without](https://user-images.githubusercontent.com/78530163/108418888-191a0900-7232-11eb-8130-08aa88d002cc.PNG)

Log with timeout
![with](https://user-images.githubusercontent.com/78530163/108418912-233c0780-7232-11eb-867c-de889309b73a.PNG)

Here is an example code. When you comment out "[Timeout]    2000" formatting is as expected.

test.robot
```
*** Settings ***
Library    MyTest.py

*** Keywords ***
My Keyword
    [Arguments]    ${text}
    [Timeout]    2000

    Log     ${text}

    MyTest.My Lib Keyword

My Other Keyword
    Log     "Log from other keyword"

*** Test cases ***
Main test
    My Keyword      "Text"
```

MyTest.py
```
from robot.api import logger
from robot.libraries.BuiltIn import BuiltIn

class MyTest():
    def my_lib_keyword(self):
        logger.info("Log from my lib keyword")
        BuiltIn().run_keyword("My Other Keyword")
```

Calling robot with
```
robot --loglevel TRACE:TRACE test.robot
```

Information about my system:

- Robot Framework 3.2.2 (Python 3.6.8 on linux)
- CentOS Linux release 7.8.2003 (Core)

**Comment by pekkaklarck** [comment]
Could you test this with RF 4.0 beta 3? After #2086 being fixed, the relative order of messages and keywords is preserved and it ought to cover also this case.

**Comment by michalxkochel** [comment]
Bad news. It looks like the problem still exists.

I got https://github.com/robotframework/robotframework/releases/tag/v4.0b3

I ran this on a bit different system (but it shouldn't be the problem):
- Robot Framework 4.0b3 (Python 3.8.5 on linux)
- Ubuntu 20.04.2 LTS

Results are:

With timeout:
![with_timeout](https://user-images.githubusercontent.com/78530163/108489789-456d6e00-72a2-11eb-9fde-46eb7d598e40.png)

Without timeout:
![without_timeout](https://user-images.githubusercontent.com/78530163/108489915-6afa7780-72a2-11eb-974b-d3a1d6b8c6c7.png)

**Comment by pekkaklarck** [comment]
There are improvements to the original like the trace log message about arguments being in correct place, but it's clear there still are problems. The reason for these problems is the combination of timeouts and using `BuiltIn.run_keyword` in a library. Why it causes problems is somewhat complicated:

- When timeouts are enabled, logging is delayed to prevent output.xml being corrupted in case timeout would occur right when the message is logged. For more details about this see #2839.
- Timeouts are in general active only when running library keywords, not when running user keywords.
- `Run Keyword` and other keywords executing other keywords are special. When Robot uses them, timeouts aren't active.
- When you use `BuiltIn.run_keyword` in your own library keyword, it doesn't detect that timeouts should be inactivated. Thus message logging, and writing `<msg>`  elements to output.xml, is delayed.
- When keywords are executed by `BuiltIn.run_keyword`, their starts and ends are written to output.xml immediately. As the result `<msg>` elements go to wrong place. If you are unlucky, timeout occurring when keyword information is written to output.xml may corrupt it.

In other words, you are using special `BuiltIn.run_keyword` functionality without Robot knowing about that. In addition to timeouts not being handled correctly, there could also be problems with arguments resolved multiple times. What you can do to avoid these problems:

1. Don't use `BuiltIn.run_keyword` if you don't really need it. In this particular example there's no real need, but I know it is a simplified example and you may actually have some concrete need for that.
2. Tell Robot that you are using `BuiltIn.run_keyword`. We have an API for that, but unfortunately the API is horrible and not really public. We have issue #2190 about cleaning it up and this issue makes it clear we definitely should get that done at some point.

**Comment by pekkaklarck** [comment]
The good news is that although the aforementioned API to tell Robot about `BuiltIn.run_keyword` usage is bad, it can be used and it seems to work fine in this case. I tested that changing the library to this works:

```python
from robot.api import logger
from robot.libraries.BuiltIn import BuiltIn, register_run_keyword

class MyTest():
    def my_lib_keyword(self, arg):
        logger.info("Log from my lib keyword with '%s'" % arg)
        BuiltIn().run_keyword("My Other Keyword", arg)


register_run_keyword('MyTest', 'my_lib_keyword', 0, deprecation_warning=False)
```
I recommend you study the documentation of that `register_run_keyword` method for more info.

**Comment by michalxkochel** [comment]
With your correction it works as I want.
I knew about `register_run_keyword`, but I thought it was optional. 
My bad I should have read the documentation.
Thanks for your comments and help.

**Comment by pekkaklarck** [comment]
Great that it works! Registering isn't mandatory but several things work better if you do that. Docs probably don't even mention timeouts. I'll leave this issue open until I add something about that. Otherwise it's enough we have #2190.

**Comment by pekkaklarck** [comment]
I noticed this same problem again and because I had forgotten about this issue reported it as #5423. The problem has now been fixed and the fix will be part of RF 7.3. The order of messages will be correct even if `register_run_keyword` isn't used. Even more importantly, timeouts won't cause problems anymore either (#5417).
